### ISSUE TITLE
Two questions about the library

### ISSUE THREAD

**Issue by RuABraun** [google/XNNPACK#148]
1. It seems the F32 GEMM implementation quantizes the input and output? I got that from the usage [here](https://github.com/google/XNNPACK/blob/master/test/fully-connected-operator-tester.h), where one has to pass min/max values of the output. I'm worried that the approximation will degrade accuracy significantly (I'm already quantising all the layers I can to int8), still have to test that, but just to confirm is there no SGEMM implementation that doesn't quantise the input and output?

2. The readme says 

> XNNPACK is a highly optimized library of **floating-point** neural network inference operators

however in the code there seems to be implementation for GEMM with int8 weights etc. ? I'm using QNNPACK at the moment for that, would it make sense to switch to XNNPACK for int8 layers?

**Comment by Maratyszcza** [comment]
1. F32 functions don't quantize computations. However, many F32 operators accept `output_min` and `output_max` arguments which enable clamping output to arbitrary range (helpful e.g. for fusing simple activation functions). If you don't want to clamp output, just set them to `+-std::numeric_limits<float>::infinity()`
2. XNNPACK is a fork of QNNPACK. Q8 functions in XNNPACK are remnants of QNNPACK code, but many internal optimizations were removed, and performance would be worse than in QNNPACK.

**Comment by RuABraun** [comment]
Okay thank you!

**Comment by RuABraun** [comment]
@Maratyszcza one more question, is the kernel shaped `(input_dim, output_dim)` or `(output_dim, input_dim)` ?

**Comment by Maratyszcza** [comment]
For FullyConnected operator, it is `(output_dim, input_dim)`. Generally operators use NHWC layout.

**Comment by RuABraun** [comment]
Works great!

Two more questions, any plans on adding an elementwise product operator? And is it somehow possible to make the FullyConnected add to the the output instead of setting it?

**Comment by Maratyszcza** [comment]
Elementwise product (including broadcasting support) landed just yesterday, see `xnn_create_multiply_nd_f32` and `xnn_setup_multiply_nd_f32`. Currently Fully Connected operator doesn't support fused addition, but you can use separate Add operator (see `xnn_create_add_nc_f32` and `xnn_setup_add_nc_f32`)

**Comment by RuABraun** [comment]
Haha perfect timing! Thanks

**Comment by Maratyszcza** [comment]
For batch normalization, I'd recommend to convert it into 1x1 depthwise convolution. XNNPACK has special optimizations for 1x1 DW convolution.

**Comment by RuABraun** [comment]
I'm not working with images, so my input/output matrixes are only 2D, I assume `xnn_create_convolution2d...` would crash then (don't see a 1D version)?

**Comment by Maratyszcza** [comment]
You can set one of the all height dimensions (input height, kernel height, height subsampling) to 1, this would be equivalent to 1D convolution.

**Comment by RuABraun** [comment]
Ah thanks, forgot to check the setup function.
